man/codlic.1
.TH codlic 1 "20 July, 2018" "0.1.4" "codlic man page"
.SH NAME
codlic \- A program for adding license text to code
.SH SYNOPSIS
codlic [options] [files]
.SH DESCRIPTION
codlic is a program that is meant for quickly licensing code files. You can
write them all, then license them all at once. It is written in Common Lisp.

Running codlic requires a license, a way to select files, and a way to select
the comment type to use.

You can specify the license with --license-file or --license-name. Using
--license-name you can specify one of the included licenses such as gplv3 or
mit. Note, these licenses are templates. The relevant information needs to be
filled in with the --license-replace option. This option takes a colon-separated
list of pairs of string separated by "/". You can escape these with a backslash.
The license defaults to MIT. To license using the MIT license with the correct
information, use "--license-name mit --license-replace '<copyright
year>/2017:<copyright holder>/Your Name'". To see the license before using it,
pass the --print-license option.

To determine what comment style to use, one of the easiest options to use is
--auto-detect-comment-type which does a lookup on some preset filetypes. This
option is used by default. You can specify individual comment languages with the
--comment-language option to specify one of the presets, such as c or
lisp.  You can also use custom comment types, such as python style with
--single-comment-string '#' or C with --opening-comment-string "/*",
--closing-comment-string " */", and --continuation-comment-string " *".
Notice the leading spaces in the closing and continuation comment strings, this
is so the formatting lines up.

To choose what files to license, simply pass files as arguments or directories
with some selection options. One way to filter out files is to use the
--filetype-language with a preset such as c or lisp. Another way is to use a
regex. To license files that end in both .c and .cc, you can use
--filetype-regex '.*\\.c|.*\\.cc' or '.*\\.cc?'. By default, codlic attempts to
license every file that is passed to it, but this will probably fail when
licensing a directory when it cannot figure out what comment type to use.
.SH OPTIONS
.IP "--license-name"
Use one of the built in licenses, options are gplv3 and mit. These require you
to edit them for information such as copyright holder, though. This means you
should copy these from the installation directory's data directory, editing
them, and using --license-file. That, or use a recursive sed replacement
command.
.IP "--license-file"
Use the given file as the header to add to code.
.IP "--filetype-language"
Only license files with the given language. Options include c and lisp.
.IP "--filetype-regex"
Only license files that match the given perl-compatible regular expression. For
example, to license all C and C++ files, use ".*\\.c|.*\\.cc" or ".*\\.cc?".
.IP "--comment-language"
License files with the comment type associated with the given language. Options
include c and lisp.
.IP "--auto-detect-comment-type"
Instead of manually specifying how to license the file, attempt to detect the
filetype and use the corresponding comment type accordingly. Currently, it
includes c and lisp.
.IP "--single-comment-string"
The string to put at the beginning of each to comment it. This option can be
used on its own to license files in the Python or Lisp style.
.IP "--opening-comment-string"
What to put at the start of the first line of a block comment for a license. Use
with --closing-comment-string and --continuation-comment-string.
.IP "--closing-comment-string"
What to put at the start of the last line of a block comment for a license. Use
with --opening-comment-string and --continuation-comment-string.
.IP "--continuation-comment-string"
What to put at the start of a line in the middle of a block comment for a
license. Use with --opening-comment-string and --closing-comment-string.
.IP "--non-blank-first-line"
Don't put a blank line with opening-comment-string when using a block comment
style. For example, don't start a file with a line that is just "/*" in C files.
.IP "--skip-file-on-error"
Don't end the licensing process when encountering an error licensing a specific
file.
.IP "--skip-shebang"
If the line starts with a shebang, e.g. "#!/bin/bash", leave that as the first
line, add a blank line, then the license.
.IP "--print-license"
Print the text of the license that would be used in a licensing. You can use
this to find out what to find and replace in the license.
.IP "--license-replace"
Do a search and replace on each line in the license file before using it. The
string passed to this argument is of the form
"search_string1/replace_string1:search_string2/replace_string2". It is a
colon-separated list of search-replace pairs separated by a forward slash.
.IP "--print-languages"
Print the available languages and their filetype regexes. Each listed language
has a corresponding comment type.
.IP "--print-licenses"
Print a list of built-in licenses and the path to their files on the filesystem.
.SH EXAMPLES

License all C files in a directory with the MIT License:

.nf
codlic --license-name mit --filetype-language c --comment-language c \\
.br
	--license-replace "<copyright year>/2017:<copyright holder> Jason Waataja" \\
.br
	my_dir
.fi

License all C++ files in a directory with the gplv3.

.nf
cp /usr/local/share/codlic/license/gplv3 .
.br
sed -i 's/Foobar/your-project/g' gplv3
.br
codlic --license-file ./gplv3 --filetype-regex '.*\\.cc' \\
.br
	--auto-detect-comment-type my_dir
.fi
.SH BUGS
There's a lot of features I know would make it work much better. I might add
them some day.
.SH AUTHOR
Jason Waataja <jasonswaataja@gmail.com>


CHANGELOG.md
# Change Log
This project adheres to Semantic Versioning

## [0.1.4] - 2018-07-20
### Changed
- Completely change build system. It now actually works, adds dependency on
  `buildapp`.
- Make some error messages include file names.

## [0.1.3] - 2017-06-12
### Changed
- Change directory structure, now has a src dir.

### Fixed
- Fix bug where it would crash when trying to run regex on pathname object
  instead of string.
- Fix mixup of closing and contiunation comment strings in the C comment type so
  it would produce invalid syntax.

## [0.1.2] - 2017-06-12
### Added
- Add --print-languages and --print-licenses which print a list of useful items
  that can be used.

### Changed
- Change filetype regexes and comment types to use one internal table based on
  language name.

### Fixed
- Fix some errors in documentation.

### Added
- Add new search-replace feature for the license. This replaces the options
  --license-search and --license-replace from before. Now it can do multiple
  searches.

## [0.1.1] - 2017-06-09
### Fixed
- Fix numerous errors in README, including incorrect executable name and
  incorrect title.

## [0.1.0] - 2017-06-09
- Initial release.


src/codlic.lisp
;;;; codlic.lisp
;;;; Main file for codlic, inserts license texts at the beginning of files.

(in-package #:codlic)

(defparameter *cmd-options*
  '(("license-name" :required nil)
    ("license-file" :required nil)
    ("filetype-language" :required nil)
    ("filetype-regex" :required nil)
    ("comment-language" :required nil)
    ("auto-detect-comment-type" :none nil)
    ("single-comment-string" :required nil)
    ("opening-comment-string" :required nil)
    ("closing-comment-string" :required nil)
    ("continuation-comment-string" :required nil)
    ("non-blank-first-line" :none nil)
    ("skip-file-on-error" :none nil)
    ("skip-shebang" :none nil)
    ("print-contents" :none nil)
    ("license-replace" :required nil)
    ("print-languages" :none nil)
    ("print-licenses" :none nil)))

(defun get-license (options)
  "Gets the correct license file for the given command line options"
  (acase (name-cons options
                    :test #'equal
                    :else-form (license-error "No license, use --license-name or --license-file."))
    ("license-name" (gethash (cdr name-cons) *license-table*))
    ("license-file" (cdr name-cons))))

(defun comment-type-for-file (path)
  (loop for language being the hash-values in *languages-table*
     with path-namestring = (namestring path)
     if (regex-matches-p (language-filetype-regex language) path-namestring)
     return (language-comment-type language)
     finally (return nil)))

(defun get-composite-comment-type (options)
  (let ((opening-string (assoc "opening-comment-string" options
                               :test #'equal))
        (closing-string (assoc "closing-comment-string" options
                               :test #'equal))
        (continuation-string (assoc "continuation-comment-string" options
                                    :test #'equal)))
    (if (and opening-string closing-string continuation-string)
        (make-composite-comment-type (cdr opening-string)
                                     (cdr closing-string)
                                     (cdr continuation-string)
                                     :blank-first-line
                                     (not (assoc "non-blank-first-line"
                                                 options
                                                 :test #'equal)))
        (license-error "When using custom comment strings, --opening-comment-string, --closing-comment-string, and --continuation-comment-string are required."))))

(defun get-comment-type (file options)
  "Gets the correct comment type based on the current argumet and previously
passed arguments. This can change based on the arg if auto-detection is used."
  (acase (comment-cons options
                       :else-form (get-composite-comment-type options)
                       :test #'equal)
    ("comment-language" (get-comment-type-for-language (cdr comment-cons)))
    ("auto-detect-comment-type" (comment-type-for-file file))
    ("single-comment-string" (make-single-comment-type (cdr comment-cons)))))

(defun should-license-p (file options)
  ;; If none of these options are present, license anything by default.
  (acase (license-cons options :else-form t :test #'equal)
    ("filetype-language" (file-matches-language-p file
                                                  (cdr license-cons)))
    ("filetype-regex" (regex-matches-p (cdr license-cons)
                                            file))))

(defun get-file-list (arg options)
  "Gets the list of files to operate on for the given argument and options. If
it points to a file then the list just contains that file but if it is a
directory then it recursively finds files in it."
  (if (uiop:directory-exists-p arg)
      (let ((file-list '()))
        (walk-directory arg
                        #'(lambda (entry)
                            (when (uiop:file-exists-p entry)
                              (push entry file-list))))
        (nreverse file-list))
      (list arg)))

(defun has-shebang (first-line)
  "Returns whether or not FIRST-LINE begins with \"#!\""
  (has-prefix first-line "#!"))

(defun create-output-lines (input-lines license-lines comment-type opts)
  "Returns the array of lines to write to the file starting with the given
license lines, then a blank line, and finally the set of input lines. Both
INPUT-LINES and LICENSE-LINES are arrays and the return value is an array."
  (if (and (assoc-equal "skip-shebang" opts)
           (plusp (length input-lines))
           (has-shebang (aref input-lines 0)))
      (let ((first-line (aref input-lines 0))
            (rest-lines (subseq input-lines 1)))
        (concatenate 'vector
                     (list first-line)
                     (list "")
                     (funcall comment-type license-lines)
                     (list "")
                     rest-lines))
      (concatenate 'vector
                   (funcall comment-type license-lines)
                   (list "")
                   input-lines)))

(defun add-license-lines (file license-lines options)
  "Adds LICENSE-LINES to the lines in FILE and writes back to it based on
OPTIONS. May signal a `license-error'."
  (when (should-license-p file options)
    (let ((comment-type
           (fail-if-nil ((get-comment-type file options))
                        'license-error
                        :text (format nil
                                      "Failed to get comment type for file: ~a."
                                      (namestring file))))
          (input-lines
           (fail-if-nil ((read-file-lines file))
                        'license-error
                        :text (format nil
                                      "Failed to read input file: ~a."
                                      (namestring file)))))
      (fail-if-nil ((write-file-lines file
                                      (create-output-lines
                                       input-lines
                                       license-lines
                                       comment-type
                                       options)))
                   'license-error
                   :text (format nil
                                 "Failed to write to file: ~a."
                                 (namestring file))))))

(defun read-license-lines (license-file opts)
  "Essentially calls READ-FILE-LINES to read the lines of LICENSE-FILE and
return them. Also makes the necessary string replacements based on OPTS."
  (let ((license-lines (read-file-lines license-file))
        (replace-opt (assoc-equal "license-replace" opts)))
    (when replace-opt
      (loop for replace-pair in (parse-search-replace-string (cdr replace-opt)) do
           (when (string= (car replace-pair) "")
             (license-error "Attempting to replace empty string."))
           (setf license-lines
                 (replace-lines license-lines
                                (car replace-pair)
                                (cdr replace-pair)))))
    license-lines))

(defun license-arg (arg options)
  "Licenses an argument based on the given options."
  (let ((license-file (get-license options))
        (file-list (get-file-list arg options)))
    (license-error-if-nil license-file "Couldn't deduce license file to use.")
    (license-error-if-nil file-list (format nil
                                            "Couldn't find any files to license for arg \"~a\""
                                            arg))
    (loop with license-lines = (read-license-lines license-file options)
       initially
         (license-error-if-nil license-lines
                               (format nil "Failed to read license file \"~a\""
                                       license-file))
       for file in file-list
       do (handler-case (add-license-lines file license-lines options)
            (license-error (err)
              (restart-case (file-license-error file
                                                (license-error-text err))
                (skip-file () nil)))))))

(defun skip-file (err)
  (declare (ignore err))
  (let ((restart (find-restart 'skip-file)))
    (when restart (invoke-restart restart))))

(defun make-file-license-error-handler (opts)
  (lambda (err)
    (let ((restart (find-restart 'skip-file)))
      (when (and restart (assoc "skip-file-on-error"
                                opts
                                :test #'equal))
        (format *error-output*
                "Warning, failed to license file \"~a\": ~a~%"
                (file-license-error-file err)
                (license-error-text err))))))

(defun check-has-license (opts)
  "Returns new version of OPTS, is destructive."
  (when (and (assoc-equal "license-name" opts)
             (assoc-equal "license-file" opts))
    (license-error "Cannot use both --license-name and --license-file"))
  (when (and (not (assoc-equal "license-name" opts))
             (not (assoc-equal "license-file" opts)))
    (setf opts (acons "license-name" "mit" opts))
    (format t "No license, using --license-name mit~%"))
  opts)

(defun check-has-filetype (opts)
  "Returns new version of OPTS, is destructive."
  (when (and (assoc-equal "filetype-language" opts)
             (assoc-equal "filetype-regex" opts))
    (license-error "Cannot use both --filetype-language and --filetype-regex"))
  (when (and (not (assoc-equal "filetype-language" opts))
             (not (assoc-equal "filetype-regex" opts)))
    (format t "No filetype, attempting to license every file.~%"))
  opts)

(defun check-has-comment-type (opts)
  "Returns new value of OPTS, is destructive."
  (exactly-one (nil
                (progn (setf opts
                             (acons "auto-detect-comment-type" nil opts))
                       (format t "No comment type, using --auto-detect-comment-type~%"))
                (license-error "Must use no more than one of --single-comment-string, composite comment options, or --auto-detect-comment-type"))
               (assoc-equal "single-comment-string" opts)
               (or (assoc-equal "opening-comment-string" opts)
                   (assoc-equal "closing-comment-string" opts)
                   (assoc-equal "continuation-comment-string" opts))
               (assoc-equal "auto-detect-comment-type" opts))
  opts)

(defun verify-options (opts)
  "Makes sure OPTS are consistent. Returns correct options if possible"
  (setf opts (check-has-license opts))
  (setf opts (check-has-filetype opts))
  (setf opts (check-has-comment-type opts))
  opts)

(defun print-license-file (license-file)
  "Returns T on success, NIL on failure."
  (let ((license-lines (read-file-lines license-file)))
    (if license-lines
        (loop for line across license-lines
           do (format t "~a~%" line)
           finally (return t))
        nil)))

(defun print-license (opts)
  "Returns T on success, NIL on failure."
  (let ((license (handler-case (get-license opts)
                   (license-error (err)
                     (format *error-output* "Cannot print license:~%~a~%"
                             (license-error-text err))
                     nil))))
    (if license
        (print-license-file license)
        nil)))

(defun print-languages ()
  "Prints each language and its filetype regex."
  (loop for language-name being the hash-keys in *languages-table*
     using (hash-value language)
     do (format t "~a: ~a~%" language-name (language-filetype-regex language))))

(defun print-licenses ()
  "Prints each license and the path to find it."
  (loop for license-name being the hash-keys in *license-table*
     using (hash-value license-path)
     do (format t "~a: ~a~%" license-name license-path)))

(defun process-args (remaining-args opts)
  "Call after reading the options. For each argument in REMAINING-ARGS, attempt
to license the file or files that it points to. The list, OPTS, is the alist of
arguments and their values."
  (loop initially
       (when (assoc-equal "print-contents" opts)
         (return (print-license opts)))
       (when (assoc-equal "print-languages" opts)
         (print-languages)
         (return t))
       (when (assoc-equal "print-licenses" opts)
         (print-licenses)
         (return t))
       (handler-case (setf opts (verify-options opts))
         (license-error (err)
           (format *error-output* "Invalid arguments.~%~a~%"
                   (license-error-text err))
           (return nil)))
       (when (null remaining-args)
         (format *error-output* "No files to license, exiting.~%")
         (return nil))
     for arg in remaining-args do
       (handler-case
           (handler-bind ((file-license-error (make-file-license-error-handler opts)))
             (license-arg arg opts))
         (license-error (err) (format *error-output*
                                      "Failure while processing argument \"~a\"~%~
~a~%Aborting.~%"
                                      arg
                                      (license-error-text err))
                        (return nil)))
     finally (return t)))
